File: new_frontend/src/types/mypage.ts
/**
 * MyPage Types
 * 마이페이지 관련 타입 정의
 *
 * Backend API와 일치하도록 설계된 타입 정의
 * Backend API: /Users/apple/Coding/ai-camp-1st-llm-agent-service-project-mockinjay/backend/app/api/mypage.py
 */

/**
 * User Profile Data
 * 사용자 프로필 정보
 *
 * Backend Response: UserProfileResponse
 */
export interface UserProfile {
  id: string;
  username: string;
  email: string;
  fullName?: string;
  bio?: string;
  profileImage?: string; // Backend primary field
  profileImageUrl?: string; // Alias for frontend compatibility (backend also returns this)
  profile: 'general' | 'patient' | 'researcher';
  role: string;
  createdAt: string; // ISO 8601 datetime string
}

/**
 * Health Profile Data
 * 건강 프로필 정보
 *
 * Backend Response: HealthProfileResponse
 */
export interface HealthProfile {
  userId: string;
  conditions: string[]; // Backend primary field
  healthConditions: string[]; // Alias for frontend compatibility (backend also returns this)
  allergies: string[];
  dietaryRestrictions: string[];
  age?: number;
  gender?: 'male' | 'female' | 'other' | 'prefer_not_to_say';
  medications?: string[]; // Frontend extended field
  notes?: string; // Frontend extended field
  updatedAt?: string; // ISO 8601 datetime string
}

/**
 * User Preferences
 * 사용자 설정
 *
 * Backend Response: UserPreferencesResponse
 */
export interface UserPreferences {
  userId: string;
  theme: 'light' | 'dark' | 'system';
  language: 'ko' | 'en';
  notifications: {
    email: boolean;
    push: boolean;
    community: boolean;
    trends: boolean; // Backend uses 'trends' instead of 'healthTips' or 'paperUpdates'
    healthTips?: boolean; // Frontend compatibility
    paperUpdates?: boolean; // Frontend compatibility
  };
  updatedAt?: string; // ISO 8601 datetime string
}

/**
 * Paper Data
 * 논문 메타데이터 (Backend의 paperData 구조)
 */
export interface PaperData {
  title: string;
  authors: string[];
  abstract: string;
  pub_date: string;
  journal?: string;
  [key: string]: any; // Allow additional fields
}

/**
 * Bookmarked Paper
 * 북마크한 논문
 *
 * Backend Response: BookmarkResponse
 * Note: Backend stores paper info in paperData, but frontend components also
 * access direct properties for convenience. Both patterns are supported.
 */
export interface BookmarkedPaper {
  id: string;
  userId: string;
  paperId: string; // PMID or unique identifier
  paperData?: PaperData; // Backend stores paper metadata as nested object
  createdAt: string; // ISO 8601 datetime string
  // Direct access properties (populated from paperData or set directly)
  title?: string;
  authors?: string[];
  abstract?: string;
  pubDate?: string;
  pub_date?: string; // Alternative field name
  journal?: string;
  url?: string;
  tags?: string[];
  notes?: string;
  bookmarkedAt?: string; // Alias for createdAt
}

/**
 * Bookmarks Response
 * 북마크 목록 응답 (with pagination)
 *
 * Backend Response: GET /api/mypage/bookmarks
 */
export interface BookmarksResponse {
  bookmarks: BookmarkedPaper[];
  total: number;
  limit: number;
  offset: number;
  hasMore: boolean;
}

/**
 * Post Data
 * 게시글 정보
 *
 * Backend Response: Posts from MongoDB
 */
export interface Post {
  id: string;
  userId: string;
  authorName: string;
  title: string;
  content: string;
  postType: 'BOARD' | 'CHALLENGE' | 'SURVEY';
  imageUrls?: string[];
  thumbnailUrl?: string;
  likes: number;
  commentCount: number;
  viewCount: number;
  // Aliases for frontend compatibility
  comments?: number; // Alias for commentCount
  views?: number; // Alias for viewCount
  createdAt: string;
  updatedAt: string;
  lastActivityAt: string;
  isPinned: boolean;
  isDeleted: boolean;
}

/**
 * Posts Response
 * 게시글 목록 응답 (with pagination)
 *
 * Backend Response: GET /api/mypage/posts
 */
export interface PostsResponse {
  posts: Post[];
  total: number;
  limit: number;
  offset: number;
  hasMore: boolean;
}

/**
 * User Level Data
 * 사용자 레벨 정보
 *
 * Backend Response: UserLevelResponse
 */
export interface UserLevelData {
  userId: string;
  level: number; // 1-5
  currentXp: number;
  requiredXp: number;
  title: string; // e.g., '새싹', '초보', '중급', '고수', '전문가'
  nextLevelTitle?: string;
  badges: Badge[];
  updatedAt?: string; // ISO 8601 datetime string
}

/**
 * Badge
 * 배지 정보
 */
export interface Badge {
  id: string;
  name: string;
  description: string;
  icon: string;
  earnedAt: string; // ISO 8601 datetime string
}

/**
 * Points Data
 * 포인트 정보
 *
 * Backend Response: PointsDataResponse
 */
export interface PointsData {
  userId: string;
  totalPoints: number;
  availablePoints: number;
  usedPoints: number;
  updatedAt?: string; // ISO 8601 datetime string
}

/**
 * Points History Item
 * 포인트 히스토리 항목
 *
 * Backend Response: PointsHistoryItemResponse
 */
export interface PointsHistoryItem {
  id: string;
  userId: string;
  amount: number; // Positive for earn, negative for spend
  type: 'earn' | 'spend' | 'expire';
  source: string; // e.g., 'quiz_completion', 'daily_login'
  description: string;
  createdAt: string; // ISO 8601 datetime string
}

/**
 * Points History Response
 * 포인트 히스토리 응답 (with pagination)
 *
 * Backend Response: PointsHistoryResponse
 */
export interface PointsHistoryResponse {
  history: PointsHistoryItem[];
  total: number;
  limit: number;
  offset: number;
  hasMore: boolean;
}

/**
 * User Notification
 * 사용자 알림
 */
export interface UserNotification {
  id: string;
  userId: string;
  type: 'comment' | 'like' | 'follow' | 'system' | 'paper_update';
  title: string;
  message: string;
  link?: string;
  read: boolean;
  createdAt: string;
}

// ============================================================================
// API Request Types
// ============================================================================

/**
 * Profile Update Request
 * 프로필 업데이트 요청
 *
 * Backend Request: UserProfileUpdateRequest
 */
export interface ProfileUpdateRequest {
  fullName?: string;
  bio?: string;
  profileImage?: string; // Backend uses 'profileImage'
}

/**
 * Health Profile Update Request
 * 건강 프로필 업데이트 요청
 *
 * Backend Request: HealthProfileUpdateRequest
 */
export interface HealthProfileUpdateRequest {
  conditions?: string[]; // Backend uses 'conditions'
  healthConditions?: string[]; // Alias for frontend compatibility
  allergies?: string[];
  dietaryRestrictions?: string[];
  age?: number; // 1-150
  gender?: 'male' | 'female' | 'other' | 'prefer_not_to_say';
  medications?: string[]; // Frontend extended field (may not be saved to backend)
  notes?: string; // Frontend extended field (may not be saved to backend)
}

/**
 * Preferences Update Request
 * 설정 업데이트 요청
 *
 * Backend Request: UserPreferencesUpdateRequest
 */
export interface PreferencesUpdateRequest {
  theme?: 'light' | 'dark' | 'system';
  language?: 'ko' | 'en';
  notifications?: {
    email?: boolean;
    push?: boolean;
    community?: boolean;
    trends?: boolean; // Backend uses 'trends'
    healthTips?: boolean; // Frontend compatibility
    paperUpdates?: boolean; // Frontend compatibility
  };
}

/**
 * Bookmark Create Request
 * 북마크 생성 요청
 *
 * Backend Request: BookmarkCreateRequest
 */
export interface BookmarkCreateRequest {
  paperId: string;
  paperData: PaperData; // Complete paper metadata
}

/**
 * Points History Filters
 * 포인트 히스토리 필터 (query parameters)
 */
export interface PointsHistoryFilters {
  limit?: number; // Default: 20, max: 50
  offset?: number; // Default: 0
  type_filter?: 'earn' | 'spend' | 'expire';
  source_filter?: string;
}

// ============================================================================
// Additional Types (Not directly from backend but useful for frontend)
// ============================================================================

/**
 * Notification Settings
 * 알림 설정 (Extended notification preferences)
 */
export interface NotificationSettings {
  userId: string;
  email: {
    enabled: boolean;
    frequency: 'instant' | 'daily' | 'weekly';
  };
  push: {
    enabled: boolean;
    sound: boolean;
    vibration: boolean;
  };
  types: {
    comments: boolean;
    likes: boolean;
    follows: boolean;
    system: boolean;
    paperUpdates: boolean;
    healthTips: boolean;
  };
  quietHours: {
    enabled: boolean;
    startTime: string; // HH:mm format
    endTime: string;   // HH:mm format
  };
  updatedAt?: string;
}

/**
 * Notification Settings Update Request
 * 알림 설정 업데이트 요청
 */
export interface NotificationSettingsUpdateRequest {
  email?: {
    enabled?: boolean;
    frequency?: 'instant' | 'daily' | 'weekly';
  };
  push?: {
    enabled?: boolean;
    sound?: boolean;
    vibration?: boolean;
  };
  types?: {
    comments?: boolean;
    likes?: boolean;
    follows?: boolean;
    system?: boolean;
    paperUpdates?: boolean;
    healthTips?: boolean;
  };
  quietHours?: {
    enabled?: boolean;
    startTime?: string;
    endTime?: string;
  };
}

// ============================================================================
// Constants
// ============================================================================

/**
 * Available health conditions
 * 선택 가능한 건강 상태
 */
export const HEALTH_CONDITIONS = [
  '당뇨병',
  '고혈압',
  '신장질환',
  '심장질환',
  '갑상선질환',
  '천식',
  '관절염',
  '고지혈증',
  '골다공증',
  '암',
  '기타',
] as const;

/**
 * Available dietary restrictions
 * 선택 가능한 식이 제한
 */
export const DIETARY_RESTRICTIONS = [
  '채식주의',
  '비건',
  '저염식',
  '저당식',
  '저지방식',
  '글루텐프리',
  '유제품 제한',
  '기타',
] as const;

/**
 * Points earned by action type (from backend)
 * 행동별 포인트 획득량
 */
export const POINTS_BY_ACTION = {
  quiz_completion: 10,
  daily_login: 5,
  community_post: 15,
  community_comment: 5,
  community_like_received: 2,
  bookmark_paper: 3,
  diet_log: 10,
  health_check: 5,
} as const;

/**
 * Level configuration (from backend)
 * 레벨 시스템 설정
 */
export const LEVEL_CONFIG = [
  { level: 1, name: '새싹', min_xp: 0, max_xp: 100 },
  { level: 2, name: '초보', min_xp: 100, max_xp: 300 },
  { level: 3, name: '중급', min_xp: 300, max_xp: 600 },
  { level: 4, name: '고수', min_xp: 600, max_xp: 1000 },
  { level: 5, name: '전문가', min_xp: 1000, max_xp: null }, // Max level
] as const;

File: backend/app/models/mypage.py
"""
MyPage Pydantic Models
Request/Response models for MyPage API endpoints
마이페이지 API 엔드포인트를 위한 요청/응답 모델
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime


# ============================================================================
# User Profile Models
# ============================================================================

class UserProfileResponse(BaseModel):
    """User profile response model (사용자 프로필 응답 모델)"""
    id: str = Field(..., description="User ID")
    username: str = Field(..., description="Username (아이디)")
    email: str = Field(..., description="Email address (이메일)")
    fullName: Optional[str] = Field(None, description="Full name (이름)")
    bio: Optional[str] = Field(None, description="User bio/description (소개)")
    profileImage: Optional[str] = Field(None, description="Profile image URL (프로필 이미지 URL)")
    profileImageUrl: Optional[str] = Field(None, description="Alias for profileImage (Frontend compatibility)")
    profile: str = Field(..., description="Profile type: general/patient/researcher")
    role: str = Field(..., description="User role: user/admin")
    createdAt: datetime = Field(..., description="Account creation date (가입일)")

    class Config:
        json_schema_extra = {
            "example": {
                "id": "507f1f77bcf86cd799439011",
                "username": "testuser",
                "email": "test@example.com",
                "fullName": "홍길동",
                "bio": "건강한 삶을 추구합니다",
                "profileImage": "/uploads/profile_123.jpg",
                "profile": "patient",
                "role": "user",
                "createdAt": "2024-01-15T10:30:00Z"
            }
        }


class UserProfileUpdateRequest(BaseModel):
    """User profile update request model (사용자 프로필 수정 요청 모델)"""
    fullName: Optional[str] = Field(None, min_length=1, max_length=100, description="Full name")
    bio: Optional[str] = Field(None, max_length=500, description="User bio/description")
    profileImage: Optional[str] = Field(None, description="Profile image URL")

    class Config:
        json_schema_extra = {
            "example": {
                "fullName": "홍길동",
                "bio": "건강한 삶을 추구하는 환자입니다",
                "profileImage": "/uploads/profile_123.jpg"
            }
        }


# ============================================================================
# Health Profile Models
# ============================================================================

class HealthProfileResponse(BaseModel):
    """Health profile response model (건강 프로필 응답 모델)"""
    userId: str = Field(..., description="User ID")
    conditions: List[str] = Field(default=[], description="Health conditions (질환)")
    healthConditions: List[str] = Field(default=[], description="Alias for conditions (Frontend compatibility)")
    allergies: List[str] = Field(default=[], description="Allergies (알레르기)")
    dietaryRestrictions: List[str] = Field(default=[], description="Dietary restrictions (식이 제한)")
    age: Optional[int] = Field(None, description="Age (나이)")
    gender: Optional[str] = Field(None, description="Gender (성별): male/female/other")
    updatedAt: Optional[datetime] = Field(None, description="Last update timestamp")

    class Config:
        json_schema_extra = {
            "example": {
                "userId": "507f1f77bcf86cd799439011",
                "conditions": ["고혈압", "당뇨"],
                "allergies": ["땅콩", "새우"],
                "dietaryRestrictions": ["저염식", "저당식"],
                "age": 45,
                "gender": "male",
                "updatedAt": "2024-01-20T14:30:00Z"
            }
        }


class HealthProfileUpdateRequest(BaseModel):
    """Health profile update request model (건강 프로필 수정 요청 모델)"""
    conditions: Optional[List[str]] = Field(None, description="Health conditions")
    allergies: Optional[List[str]] = Field(None, description="Allergies")
    dietaryRestrictions: Optional[List[str]] = Field(None, description="Dietary restrictions")
    age: Optional[int] = Field(None, ge=1, le=150, description="Age (1-150)")
    gender: Optional[str] = Field(None, description="Gender: male/female/other")

    class Config:
        json_schema_extra = {
            "example": {
                "conditions": ["고혈압", "당뇨"],
                "allergies": ["땅콩"],
                "dietaryRestrictions": ["저염식"],
                "age": 45,
                "gender": "male"
            }
        }


# ============================================================================
# User Preferences Models
# ============================================================================

class UserPreferencesResponse(BaseModel):
    """User preferences response model (사용자 설정 응답 모델)"""
    userId: str = Field(..., description="User ID")
    theme: str = Field(default="light", description="UI theme: light/dark")
    language: str = Field(default="ko", description="Language: ko/en")
    notifications: Dict[str, bool] = Field(
        default={
            "email": True,
            "push": True,
            "community": True,
            "trends": True
        },
        description="Notification settings"
    )
    updatedAt: Optional[datetime] = Field(None, description="Last update timestamp")

    class Config:
        json_schema_extra = {
            "example": {
                "userId": "507f1f77bcf86cd799439011",
                "theme": "light",
                "language": "ko",
                "notifications": {
                    "email": True,
                    "push": True,
                    "community": True,
                    "trends": True
                },
                "updatedAt": "2024-01-20T14:30:00Z"
            }
        }


class UserPreferencesUpdateRequest(BaseModel):
    """User preferences update request model (사용자 설정 수정 요청 모델)"""
    theme: Optional[str] = Field(None, description="UI theme: light/dark")
    language: Optional[str] = Field(None, description="Language: ko/en")
    notifications: Optional[Dict[str, bool]] = Field(None, description="Notification settings")

    class Config:
        json_schema_extra = {
            "example": {
                "theme": "dark",
                "language": "en",
                "notifications": {
                    "email": True,
                    "push": False,
                    "community": True,
                    "trends": False
                }
            }
        }


# ============================================================================
# Bookmark Models
# ============================================================================

class BookmarkResponse(BaseModel):
    """Bookmark response model (북마크 응답 모델)"""
    id: str = Field(..., description="Bookmark ID")
    userId: str = Field(..., description="User ID")
    paperId: str = Field(..., description="Paper PMID or unique identifier")
    paperData: Dict[str, Any] = Field(..., description="Paper metadata (title, authors, abstract, etc.)")
    createdAt: datetime = Field(..., description="Bookmark creation timestamp")

    class Config:
        json_schema_extra = {
            "example": {
                "id": "507f1f77bcf86cd799439011",
                "userId": "507f1f77bcf86cd799439012",
                "paperId": "12345678",
                "paperData": {
                    "title": "Research on Hypertension Treatment",
                    "authors": ["Kim, J.", "Lee, S."],
                    "abstract": "This study investigates...",
                    "pub_date": "2024-01-15"
                },
                "createdAt": "2024-01-20T10:00:00Z"
            }
        }


class BookmarkCreateRequest(BaseModel):
    """Bookmark creation request model (북마크 생성 요청 모델)"""
    paperId: str = Field(..., description="Paper PMID or unique identifier")
    paperData: Dict[str, Any] = Field(..., description="Paper metadata to store")

    class Config:
        json_schema_extra = {
            "example": {
                "paperId": "12345678",
                "paperData": {
                    "title": "Research on Hypertension Treatment",
                    "authors": ["Kim, J.", "Lee, S."],
                    "abstract": "This study investigates...",
                    "pub_date": "2024-01-15",
                    "journal": "Medical Journal"
                }
            }
        }


# ============================================================================
# Points & Level Models
# ============================================================================

class UserLevelResponse(BaseModel):
    """User level response model (사용자 레벨 응답 모델)"""
    userId: str = Field(..., description="User ID")
    level: int = Field(..., ge=1, le=5, description="Current level (1-5)")
    currentXp: int = Field(..., ge=0, description="Current XP points")
    requiredXp: int = Field(..., description="XP required for current/next level")
    title: str = Field(..., description="Level title (e.g., '새싹', '전문가')")
    nextLevelTitle: Optional[str] = Field(None, description="Next level title")
    badges: List[Dict[str, Any]] = Field(default=[], description="Earned badges")
    updatedAt: Optional[datetime] = Field(None, description="Last update timestamp")

    class Config:
        json_schema_extra = {
            "example": {
                "userId": "507f1f77bcf86cd799439011",
                "level": 2,
                "currentXp": 150,
                "requiredXp": 300,
                "title": "초보",
                "nextLevelTitle": "중급",
                "badges": [
                    {"id": "first_quiz", "name": "첫 퀴즈 완료", "description": "첫 퀴즈를 완료했습니다", "icon": "trophy", "earnedAt": "2024-01-15T10:00:00Z"}
                ],
                "updatedAt": "2024-01-20T14:30:00Z"
            }
        }


class PointsDataResponse(BaseModel):
    """Points data response model (포인트 데이터 응답 모델)"""
    userId: str = Field(..., description="User ID")
    totalPoints: int = Field(..., ge=0, description="Total earned points")
    availablePoints: int = Field(..., ge=0, description="Available (spendable) points")
    usedPoints: int = Field(..., ge=0, description="Used/spent points")
    updatedAt: Optional[datetime] = Field(None, description="Last update timestamp")

    class Config:
        json_schema_extra = {
            "example": {
                "userId": "507f1f77bcf86cd799439011",
                "totalPoints": 250,
                "availablePoints": 200,
                "usedPoints": 50,
                "updatedAt": "2024-01-20T14:30:00Z"
            }
        }


class PointsHistoryItemResponse(BaseModel):
    """Points history item response model (포인트 히스토리 항목 응답 모델)"""
    id: str = Field(..., description="Transaction ID")
    userId: str = Field(..., description="User ID")
    amount: int = Field(..., description="Points amount (positive=earn, negative=spend)")
    type: str = Field(..., description="Transaction type: earn/spend/expire")
    source: str = Field(..., description="Source of points (e.g., quiz_completion)")
    description: str = Field(..., description="Human-readable description")
    createdAt: datetime = Field(..., description="Transaction timestamp")


class PointsHistoryResponse(BaseModel):
    """Points history response model (포인트 히스토리 응답 모델)"""
    history: List[PointsHistoryItemResponse] = Field(..., description="History items")
    total: int = Field(..., description="Total count of transactions")
    limit: int = Field(..., description="Page size")
    offset: int = Field(..., description="Page offset")
    hasMore: bool = Field(..., description="Whether more items exist")
